rbc/omnisci_backend/omnisci_buffer.py

```diff
@@ -387,12 +388,88 @@ def codegen(context, builder, sig, args):
     return sig, codegen
 
 
+@extending.intrinsic
+def omnisci_buffer_set_null_(typingctx, data):
+    sig = types.none(data)
+
+    def codegen(context, builder, sig, args):
+        rawptr = cgutils.alloca_once_value(builder, value=args[0])
+        ptr = builder.load(rawptr)
+        builder.store(int8_t(1), builder.gep(ptr, [int32_t(0), int32_t(2)]))
+
+    return sig, codegen
+
+
+@extending.intrinsic
+def omnisci_array_is_null_(typingctx, T, elem):
+    sig = types.boolean(T, elem)
+
+    target_info = TargetInfo()
+    null_value = target_info.null_values[str(T.dtype)]
+    nv = ir.Constant(ir.IntType(T.dtype.bitwidth), null_value)
+
+    def codegen(context, builder, signature, args):
+        _, elem = args
+        if isinstance(T.dtype, types.Float):
+            elem = builder.bitcast(elem, nv.type)
+        return builder.icmp_signed('==', elem, nv)
+
+    return sig, codegen
+
+
 # "BufferPointer.is_null" checks if a given array or column is null
 # as opposed to "BufferType.is_null" that checks if an index in a
 # column is null
 @extending.overload_method(BufferPointer, 'is_null')
-def omnisci_buffer_is_null(x):
+def omnisci_buffer_is_null(x, row_idx=None):
+    T = x.eltype
     if isinstance(x, BufferPointer):
-        def impl(x):
-            return omnisci_buffer_is_null_(x)
+        if cgutils.is_nonelike(row_idx):
+            def impl(x, row_idx=None):
+                return omnisci_buffer_is_null_(x)
+        else:
+            def impl(x, row_idx=None):
+                return omnisci_array_is_null_(T, x[row_idx])
+        return impl
+
+
+@extending.intrinsic
+def omnisci_array_set_null_(typingctx, arr, row_idx):
+    T = arr.eltype
+    sig = types.none(arr, row_idx)
+
+    target_info = TargetInfo()
+    null_value = target_info.null_values[f'Array<{T}>']
+    # it looks like that the null values for arrays are:
+    null_value = (null_value - 1) * (-1)
```

Comment:
Notice that omnisci sends the null values as unsigned integers. Hence, omnisci_array_is_null_ should use unsigned version of `icmp`.

Reply:
Another note: the null_value (in Python) is misinterpreted as a signed integer.
Using Python int (that `null_value` is here), the expression `(null_value - 1) * (-1)` is valid (because Python int has no bitwidth) but when using fixed width integer (say, numpy int dtype), it would be integer overflow.

I hope we can bring some sense to this transformation before accepting it.

Reply:
```
>>> null_value = 9223372036854775809 # 'Array<int64>'
>>> str(ir.Constant(ir.IntType(64), numpy.uint64(null_value).astype(numpy.int64)))
'i64 -9223372036854775807'
>>> str(ir.Constant(ir.IntType(64), null_value))
'i64 9223372036854775809'
>>> str(ir.Constant(ir.IntType(64), -(null_value-1)))
'i64 -9223372036854775808'
```
I would expect `i64 -9223372036854775807` to be correct rather than `i64 -9223372036854775808`.

Reply:
I figure it out what was wrong. The server is sending null values as unsigned ones rather than signed. See:
https://godbolt.org/z/37dnvd

Reply:
So,
```
>>> null_value = 9223372036854775809 # 'Array<int64>'
>>> str(ir.Constant(ir.IntType(64), numpy.uint64(null_value).astype(numpy.int64)))
'i64 -9223372036854775807'
```
corresponds to the result that you referred in https://godbolt.org/z/37dnvd , as expected.
Yes, the server sends the null values as (stringified) unsigned ints (which is not wrong, it is just a choice).
When `ir.Constant` is transformed to a LLVM IR string, we should see the null value as the corresponding signed int.
So I think the correct way to create the `ir.Constant` instance is to use:
```
  ir.Constant(ir.IntType(64), numpy.uint64(null_value).astype(numpy.int64)
```

Reply:
Except for the last part, that's what I did:
```python
T = ... # typeof null_value
nv = numpy.uint64(null_value).astype(T)
constant = ir.Constant(ir.IntType(bits), nv)
```

See:
https://github.com/xnd-project/rbc/blob/ba5a169c97b0cf8baa4cb176fd45700b033b992d/rbc/omnisci_backend/omnisci_buffer.py#L451-L456

Reply:
Shouldn't the same transform be applied to float types? E.g. `null_value = np.uint64(null_value).astype(np.int64)` for float64. Also, np.uint64(null_value) does not sound right, say, for int16 null value because `np.uint64(..).astype('int16')` looses bits. I think it should read `np.uint16(..).astype(np.int16)` for int16 null value, for instance.

Or better yet, have:
```
null_value = np.uint64(null_value)
nv = ir.Constant(ir.IntType(T.dtype.bitwidth), null_value)
# in is_null codegen:
return builder.icmp_unsigned('==', elem, nv)
# set_null code needs to be tested
```

The aim is to obtain code with fewer conditionals and fewer uint->int transformations as we just want the bit content of null values to be correct.

Reply:
> Shouldn't the same transform be applied to float types?

No. For floats the server sends the correct value. See: https://godbolt.org/z/x4of5o

> Also, np.uint64(null_value) does not sound right, say, for int16 null value because np.uint64(..).astype('int16') looses bits. I think it should read np.uint16(..).astype(np.int16) for int16 null value, for instance.

Well, I tested all the combinations on my machine and `np.uint64` seems to be converting the data correctly. Nevertheless, I can change this to `np.uT(nv).astype(T)` without any problems.

> `nv = ir.Constant(ir.IntType(T.dtype.bitwidth), null_value)`

This is not safe. llvmlite accepts any value for null_value and it will not overflow it as we want. For instance,
`ir.Constant(ir.IntType(8), 129) != -127`

Reply:
> > Shouldn't the same transform be applied to float types?
> 
> No. For floats the server sends the correct value. See: https://godbolt.org/z/x4of5o

This correctness test relies on the used printf formatting. Notice that you are printing unsigned values using signed formatting and hence the negative values.

> > Also, np.uint64(null_value) does not sound right, say, for int16 null value because np.uint64(..).astype('int16') looses bits. I think it should read np.uint16(..).astype(np.int16) for int16 null value, for instance.
> 
> Well, I tested all the combinations on my machine and `np.uint64` seems to be converting the data correctly. Nevertheless, I can change this to `np.uT(nv).astype(T)` without any problems.

It will likely matter when endianness becomes a factor.

> 
> > `nv = ir.Constant(ir.IntType(T.dtype.bitwidth), null_value)`
> 
> This is not safe. llvmlite accepts any value for null_value and it will not overflow it as we want. For instance,
> `ir.Constant(ir.IntType(8), 129) != -127`

Makes sense, although, I would not want to deal with overflow, just having the memory bits correct is sufficient..